### PR TITLE
Update class_mailhandler.php

### PR DESCRIPTION
**Pull request by NotoriousPyro**
When a user changes their password, an email is received from the board administrator. This "functionality" exposes the administrator's email to anyone who registers.

Thus, this change allows a person to set ReturnEmail to donotreply@domain.com and AdminEmail to adminemail@domain.com, the emails from Contact Us, etc will still work fine and be delivered to the administrator, yet any other emails not specifically defined as From anyone will be sent via donotreply@domain.com.

For example, the original file does this:
* A user uses the Contact Us link, the email is delivered to AdminEmail received from the user's own email address.
* A user registers, the email is delivered to the user's own email address, received from **AdminEmail**.

These changes do:
* A user uses the Contact Us link, the email is delivered to AdminEmail received from the user's own email address.
* A user registers, the email is delivered to the user's own Email Address, received from **ReturnEmail**.

### CODE REVIEW

**Comment by NotoriousPyro** [comment]
Forgot to leave a commit message for my last edit, I tidied up the code and reduced code-repeation.

**Comment by Sama34** [comment]
I do feel it is now inconsistent but if it does works.

@mybb/developers @Destroy666x @JN-Jones

**Comment by Destroy666x** [comment]
Not sure if it solves the problem yet, but llinking to a similar issue: https://github.com/mybb/mybb/issues/257

**Comment by NotoriousPyro** [comment]
This was never designed to fix that issue, it was designed to fix the issue of mail being delivered inconsistently between senders.

File: inc/class_mailhandler.php [hunk 1]
```diff
@@ -92,6 +92,23 @@ class MailHandler
 	public $parse_format = 'text';
 
 	/**
+	 * Selects between AdminEmail and ReturnEmail, dependant on if ReturnEmail is filled.
```
Comment (JN-Jones) [inline inc/class_mailhandler.php]: This should get a php doc paramater:
`@return string`

File: inc/class_mailhandler.php [hunk 2]
```diff
@@ -92,6 +92,23 @@ class MailHandler
...
+	 */
+	function from_select()
+	{
+		global $mybb;
+		
+		if($mybb->settings['returnemail'])
```
Comment (JN-Jones) [inline inc/class_mailhandler.php]: Should be checked via `empty` or at least `trim`, just in case the user entered a space.

**Comment by JN-Jones** [comment]
Except those two comments I also think that we shouldn't have a function which needs to be called like this: `$mybb->settings[$this->from_select]`. Instead that function should already return the correct email imho (and should be renamed to something like `get_from_email`).

**Comment by euantorano** [comment]
Definitely agree with @JN-Jones. I'd much prefer a `MailHandler::get_from_email() -> string` method that returns the actual from address, rather than an array key name.

**Comment by JN-Jones** [comment]
@NotoriousPyro

**Comment by NotoriousPyro** [comment]
Requested changes added, not extensively tested so proceed with caution.

**Comment by JN-Jones** [comment]
Can you also fix the Doc Block I've mentioned? Rest seems fine to me.

**Comment by NotoriousPyro** [comment]
Done.

**Comment by JN-Jones** [comment]
Thanks. @Destroy666x @euantorano if both of you are fine with this PR I'll merge it with the others.

**Comment by Destroy666x** [comment]
The current form of the PR renders the `$from` parameter used in many places in MyBB and plugins completely useless...

What should be done:
- don't touch `$this->from` if it's passed to the function
- the initial issue which could be fixed here is caused by invalid `Reply-To` header: https://github.com/NotoriousPyro/mybb/blob/patch-1/inc/class_mailhandler.php#L299 which is based on `$this->return_email` not `$this->from`

**Comment by NotoriousPyro** [comment]
Reverted.

**Comment by NotoriousPyro** [comment]
Probably the best way to check if SPF is in use is to actually ask the administrator. It could be a toggled setting in the email settings, constant DNS lookups I costly, both to PHP and DNS servers.

**Comment by JN-Jones** [comment]
IMHO #257 should be fixed seperatly anyways. Would like to merge this without any fix for it @Destroy666x @NotoriousPyro

**Comment by Destroy666x** [comment]
Yes, as I said above a new PR would be better. And an issue should be correct for this one because there is none.

**Comment by JN-Jones** [comment]
Ok, will create an issue later and merge this.

**Comment by JN-Jones** [comment]
#2054

**Comment by JN-Jones** [comment]
Thanks again for this PR!

**Pull request merged**